lib/charms/tempo_k8s/v2/tracing.py

```diff
@@ -902,9 +902,19 @@ def _get_endpoint(
     def get_endpoint(
         self, protocol: ReceiverProtocol, relation: Optional[Relation] = None
     ) -> Optional[str]:
-        """Receiver endpoint for the given protocol."""
+        """Receiver endpoint for the given protocol.
+
+        Raises:
+        ProtocolNotRequestedError:
+            If the charm unit is the leader unit and attempts to obtain an endpoint for a protocol it did not request.
+        """
         endpoint = self._get_endpoint(relation or self._relation, protocol=protocol)
         if not endpoint:
+            # non-leader units cannot access local app databag
+            if not self._charm.unit.is_leader():
+                # if its a racing condition, we can return None until the provider publishes the endpoint in a later event.
```

Comment:
```suggestion
                # if its a race condition, we can return None until the provider publishes the endpoint in a later event.
```
Also it's not clear to me what is racing with what?

Reply:
if the requirer tries to get the endpoint before the provider publishes it

Reply:
ah, ok, I don't usually think of those as race conditions but "just juju", but I guess technically they are :)